jni/src/Foreign/JNI/Types.hs

```diff
@@ -220,7 +227,7 @@ instance Storable JValue where
   poke p (JLong x) = poke (castPtr p) x
   poke p (JFloat x) = poke (castPtr p) x
   poke p (JDouble x) = poke (castPtr p) x
-  poke p (JObject x) = poke (castPtr p :: Ptr (J a)) x
+  poke p (JObject (J x)) = withForeignPtr x (poke (castPtr p))
```

Comment:
This use of the underlying `Ptr` is unsafe. The pointer is written while the reference is valid, but we don't know when it will be read. Writing a local ref would be better. Either we create the local ref here, or we tell the user that `poke` should never be used on global refs.